File: level2/ex57.py
from functools import reduce

def solution(n):
    # 삼각 배열 
    arr = [[0] * (i+1) for i in range(n)]
    # 현재 숫자, 행, 열
    number, row, column = 1, 0, 0
    # 이동 방향 (세로, 가로, 대각선)
    moves, move = [(1, 0), (0, 1), (-1, -1)], 0
    
    for i in range(n, -1, -1):
        for j in range(i):
            arr[row][column] = number
            number += 1
            move = move if j < i-1 else (move + 1) % 3
            row += moves[move][0]
            column += moves[move][1]
        
    return reduce(lambda x, y: x + y, arr)

File: level2/ex57.js
function solution(n) {
    // 삼각 배열
    const arr = Array.from({ length: n }, (_, i) => Array(i+1).fill(0));
    // 현재 숫자, 행, 열, 이동 방향
    let [number, row, column, move] = [1, 0, 0, 0];
    // 이동 방향 배열
    const moves = [[1, 0], [0, 1], [-1, -1]];
    
    for (let i = n; i >= 0; i--) {
        for (let j = 0; j < i; j++) {
            arr[row][column] = number++;
            move = j < i - 1 ? move : (move + 1) % 3;
            row += moves[move][0];
            column += moves[move][1];
        }
    }

    return arr.flatMap(v => v);
}